napalm_eos/eos.py

```diff
@@ -1201,3 +1201,140 @@ def traceroute(self, destination, source='', ttl=0, timeout=0):
+            commands.append('show ipv6 bgp neighbors vrf all')
+        else:
+            try:
+                peer_ver = IPAddress(neighbor_address).version
+            except Exception:
+                LOG.error('Invalid BGP neighbor address %s' % neighbor_address)
+                return
+
+            if peer_ver == 4:
+                commands.append('show ip bgp neighbors %s vrf all' %
+                                neighbor_address)
+            elif peer_ver == 6:
+                commands.append('show ipv6 bgp neighbors %s vrf all' %
+                                neighbor_address)
+
+        raw_output = (
+            self.device.run_commands(commands, encoding='text'))
+
+        bgp_detail_info = {}
+
+        if neighbor_address:
+            peer_info = self._parse_per_peer_bgp_detail(
+                raw_output[0]['output'])
+
+            # Appending peer_info to final data structure
+            if int(peer_info[0]['remote_as']) not in bgp_detail_info:
+                bgp_detail_info[int(peer_info[0]['remote_as'])] = []
+
+            (bgp_detail_info[int(peer_info[0]['remote_as'])].
+                append(peer_info[0]))
+
+        else:
+            # Using preset template to extract peer info
+            v4_peer_info = self._parse_per_peer_bgp_detail(
+                raw_output[0]['output'])
+
+            v6_peer_info = self._parse_per_peer_bgp_detail(
+                raw_output[1]['output'])
+
+            for peer_info in v4_peer_info:
+                # Appending v4 peer_info to final data structure
+                if int(peer_info['remote_as']) not in bgp_detail_info:
+                    bgp_detail_info[int(peer_info['remote_as'])] = []
+
+                bgp_detail_info[int(peer_info['remote_as'])].append(peer_info)
+
+            for peer_info in v6_peer_info:
+                # Appending v6 peer_info to final data structure
+                if int(peer_info['remote_as']) not in bgp_detail_info:
+                    bgp_detail_info[int(peer_info['remote_as'])] = []
+
+                bgp_detail_info[int(peer_info['remote_as'])].append(peer_info)
+
+        return bgp_detail_info
+
+    def _parse_per_peer_bgp_detail(self, peer_output):
+        """This function parses the raw data per peer and returns a
+        json structure per peer.
+        """
+
+        int_fields = ['accepted_prefix_count', 'local_as', 'remote_as',
+                      'local_port', 'remote_port', 'local_port',
+                      'input_messages', 'output_messages', 'input_updates',
+                      'output_updates', 'messages_queued_out', 'holdtime',
+                      'configured_holdtime', 'keepalive',
+                      'configured_keepalive', 'advertise_prefix_count',
+                      'received_prefix_count']
+
+        peer_details = []
+
+        v4_summary = self.device.run_commands(
+            'show ip bgp summary vrf all', encoding='json')
+
+        v6_summary = self.device.run_commands(
```

Comment:
Please remove this one too - not used and creates supplementary requests to the device.